_posts/2019-03-28-Building-With-Google-Cloud-Build.md

```diff
@@ -0,0 +1,468 @@
+---
+layout: post
+authors: [tom_verelst]
+title: 'Building with Google Cloud Build'
+image: /img/2019-03-28-cloudbuild/cloudbuild.png
+tags: [GCP,Docker,DevOps,Git,Kubernetes]
+category: Cloud
+comments: false
+---
+
+In this post, 
+I will give a quick overview on what is possible with Google Cloud Build.
+
+Google Cloud Build is a fully managed solution to building containers or other artifacts.
+It integrates with Google Storage, Cloud Source Repositories, GitHub or BitBucket.
+
+## A simple YAML file
+
+You can easily set up a build pipeline using a YAML file which you store in your source code.
+Each build step is defined using a container image and passing arguments to it.
+Here is an example:
+
+```
+steps:
+  # Test Helm templates
+  - name: 'quay.io/helmpack/chart-testing:v2.2.0'
+    id: 'Helm Lint'
+    args: ['ct', 'lint', '--all', '--chart-dirs', '/workspace/helm', '--validate-maintainers=false']
+
+  # Build image
+  - name: 'gcr.io/cloud-builders/docker'
+    id: 'Building image'
+    args: ['build', '-t', 'eu.gcr.io/$PROJECT_ID/cloud-build-demo:$COMMIT_SHA', '.']
+
+  # Create custom image tag and write to file /workspace/_TAG
+  - name: 'ubuntu'
+    id: 'Setup'
+    args: ['bash', '-c', "echo `echo $BRANCH_NAME | sed 's,/,-,g' | awk '{print tolower($0)}'`_$(date -u +%Y%m%dT%H%M)_$SHORT_SHA > _TAG; echo $(cat _TAG)"]
+
+  # Tag image with custom tag
+  - name: 'gcr.io/cloud-builders/docker'
+    id: 'Tagging image'
+    entrypoint: '/bin/bash'
+    args: ['-c', "docker tag eu.gcr.io/$PROJECT_ID/cloud-build-demo:$COMMIT_SHA eu.gcr.io/$PROJECT_ID/ms-map-report:$(cat _TAG)"]
+images: ['eu.gcr.io/$PROJECT_ID/cloud-build-demo']
+timeout: 15m
+options:
+  machineType: 'N1_HIGHCPU_8'
+```
+
+You are free to use any image that you like.
+Cloud Build already [provides a set of base images (called Cloud Builders)](https://cloud.google.com/cloud-build/docs/cloud-builders),
+including images for Maven, Git, Docker, Bazel, npm, gcloud, kubectl and so on. 
+
+You can also customise some build options like the timeout time of the build,
+or on which kind of node the build runs.
+Pricing is done based on the amount of build minutes. 
+However, 
+if you use the default node, 
+the first 120 build minutes are free every day!
+
+If the build finishes successfully,
+Cloud Build will automatically upload the built images to the container registry.
+This is based on the images defined in the `images` array.
+
+Data usually needs be shared in between steps.
+You might want to download dependencies in one step,
+and build your artifact in another step,
+or run tests in a separate step.
```

Comment:
You might want to download dependencies in step 1,
build your artifact in step 2 and run tests in step 3.